glean-core/src/histogram/functional.rs

```diff
@@ -0,0 +1,152 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+
+use super::{Bucketing, Histogram};
+
+/// A functional bucketing algorithm.
+///
+/// Bucketing is performed by a function, rather than pre-computed buckets.
+/// The bucket index of a given sample is determined with the following function:
+///
+/// i = ⌊n log<sub>base</sub>(𝑥)⌋
+///
+/// In other words, there are n buckets for each power of `base` magnitude.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Functional {
+    exponent: f64,
+}
+
+impl Functional {
+    /// Instantiate a new functional bucketing.
+    fn new(log_base: f64, buckets_per_magnitutde: f64) -> Functional {
+        let exponent = log_base.powf(1.0 / buckets_per_magnitutde);
+
+        Functional { exponent }
+    }
+
+    /// Maps a sample to a "bucket index" that it belongs in.
+    /// A "bucket index" is the consecutive integer index of each bucket, useful as a
+    /// mathematical concept, even though the internal representation is stored and
+    /// sent using the minimum value in each bucket.
+    fn sample_to_bucket_index(&self, sample: u64) -> u64 {
+        ((sample + 1) as f64).log(self.exponent) as u64
+    }
+
+    /// Determines the minimum value of a bucket, given a bucket index.
+    fn bucket_index_to_bucket_minimum(&self, index: u64) -> u64 {
+        self.exponent.powf(index as f64) as u64
+    }
+}
+
+impl Bucketing for Functional {
+    fn sample_to_bucket_minimum(&self, sample: u64) -> u64 {
+        if sample == 0 {
+            return 0;
+        }
+
+        let index = self.sample_to_bucket_index(sample);
+        self.bucket_index_to_bucket_minimum(index)
+    }
+}
+
+impl Histogram<Functional> {
+    /// Create a histogram with functional buckets.
+    pub fn functional(log_base: f64, buckets_per_magnitutde: f64) -> Histogram<Functional> {
+        Histogram {
+            values: HashMap::new(),
+            count: 0,
+            sum: 0,
+            bucketing: Functional::new(log_base, buckets_per_magnitutde),
+        }
+    }
+
+    /// Get a snapshot of all contiguous values.
+    pub fn snapshot_values(&self) -> HashMap<u64, u64> {
+        if self.values.is_empty() {
+            return HashMap::new();
+        }
+
+        let mut min_key = None;
+        let mut max_key = None;
+
+        // `Iterator#min` and `Iterator#max` would do the same job independently,
+        // but we want to avoid iterating the keys twice, so we loop ourselves.
+        for key in self.values.keys() {
+            let key = *key;
+
+            if min_key.is_none() || key < min_key.unwrap() {
+                min_key = Some(key);
+            }
+
+            if max_key.is_none() || key > max_key.unwrap() {
+                max_key = Some(key);
+            }
```

Comment:
There's potential for a bug here if we're not careful.  I might just be missing some parts of the code here, so feel free to just point me at them if I'm wrong.

This *assumes* that only non-zero-valued entries are in the map in order to find the min/max.  If you ran this over values that were already serialized/deserialized from the database (and thus have an extra zero-valued entry at the end), you would end up adding an extra element to the end each time you do so.  The glean-ac implementation gets around this by using a different serialization for (a) persistence vs. (b) the ping payload.  For (a), it doesn't insert the zero values and just saves the "raw" sparse data.  For (b) it does.  Not only does that avoid that bug, it avoids doing extra work for the local persistence.  I *think* here, it's always inserting the contiguous values for both (but I could be wrong -- that serialization is more implicit here and I just might be missing it...)  If having two different serializations doesn't make sense in this implementation, then I think we at least need to fix the bug along the lines of:

```
  for (key, val) in self.values.iter() {
      if val == 0 {
          continue;
      }
      ...
  }
```

We're also going to need to do something similar for custom distributions, where we need to serialize the parameters for persistence, but we don't need to send them in the payload...

Reply:
We serialize the `Histogram<Functional>` into the database, that is only the hash map of stored values.
As we don't store 0-buckets, we can't run into that problem.

persistence format and payload format are already different.